public/js/widget-manager.js

```diff
@@ -45,16 +61,73 @@ define(['jupyter-js-widgets'], function(Widgets) {
         });
     };
 
+    /*
+     * Returns callbacks to be invoked in response to ipywidget triggered
+     * comm messages on shell, input, and iopub channels. Each callback takes
+     * one parameter: the comm message from the kernel. The return value should
+     * be an object with any of the following properties defined as functions.
+     *
+     * iopub.status
+     * iopub.clear_output
+     * iopub.output
+     * shell.reply
+     * input
+     *
+     * view: ipywidget view instance
+     */
+    WidgetManager.prototype.callbacks = function(view) {
+        // TODO: only registering one callback until https://github.com/ipython/ipywidgets/pull/353
+        // is fixed
+        var mgr = this;
+        var callbacks = {
+            iopub : {
+                output: function(msg) {
+                    console.debug(msg.msg_type, view);
+                    // TODO: it's probably not always parent
+                    mgr.msgHandler(msg, view.options.parent.options.outputAreaModel);
```

Comment:
Look up the ancestor chain until we find the view.options that has `outputAreaModel` or the parent is null or `options.root: true` (which should be the same as `parent = null`).